// @dmi-bootcamp/pet-layer/Pet.js
const assert = require('assert');
const uuidv4 = require('uuid/v4');

class Pet {
    constructor(inputObject) {
        assert(inputObject.name, 'Pet must have a `name`');
        this.name = inputObject.name;

        assert(inputObject.species, 'Pet must have a `species`');
        this.species = inputObject.species;

        assert(inputObject.description, 'Pet must have a `description`');
        this.description = inputObject.description;
    }

    toDbJson() {
        this.id = this.id || uuidv4();

        const p = {
            id: this.id,
            name: this.name,
            species: this.species,
            description: this.description,
        };

        return p;
    }

    toUpdateExpressions() {
        return {
            UpdateExpression: 'set #name = :n, species = :s, description = :d',
            ExpressionAttributeNames: {
                '#name': 'name',
            },
            ExpressionAttributeValues: {
                ':n': this.name,
                ':s': this.species,
                ':d': this.description,
            }
        };
    }

    static fromDbJson(inputObject) {
        const pet = new Pet(inputObject);
        pet.id = inputObject.id;

        return pet;
    }
}

module.exports = Pet;
